src/algorithms/piecewise_linear_stretching.jl

```diff
@@ -0,0 +1,179 @@
+        for src_pair in src_intervals
+            if first(src_pair) == last(src_pair) 
+                throw(ArgumentError("The start and end of an interval in src_intervals cannot be equal."))
+            end
+        end
+        T = promote_type(T₁, T₂)
+        new{T}(src_intervals, dst_intervals)
+    end    
+end
+
+function (f::PiecewiseLinearStretching)(out::GenericGrayImage, img::GenericGrayImage) 
+    # Verify that the specified dst_intervals fall inside the permissible range
+    # of the output type. 
+    T = eltype(out)
+    for (a,b) in f.dst_intervals
+        p = typemin(T)
+        r = typemax(T)
+        if  !(p <= a <= r) || !(p <= b <= r)
+            throw(DomainError("The specified interval [$a, $b] falls outside the permissible range [$p, $r] associated with $T"))
+        end
+    end
+    # A linear mapping of a value x in the interval [A, B] to the interval [a,b] is given by
+    # the formula f(x) = (x-A) * ((b-a)/(B-A)) + a. The operation r * x - o is equivalent to
+    # (x-A) * ((b-a)/(B-A)) + a. Precalculate the r and o so that later on an inner loop
+    # contains only multiplication and addition to get better performance.
+    N = length(f.src_intervals)
+    rs = zeros(Float64, N)
+    os = zeros(Float64, N)
+    for (src, dst) in zip(pairs(f.src_intervals), pairs(f.dst_intervals)) 
+        n, (A,B) = src
+        _, (a,b) = dst   
+        rs[n] = (b - a) / (B - A)
+        os[n] = (A*b - B*a) / (B - A)
+    end
+
+    # Determine which interval the source pixel fall into, and apply the accompanying linear
+    # transformation taking into account NaN values. 
+    @inbounds for p in eachindex(img)
+        val = img[p]
+        if isnan(val)
+            out[p] = val
+        else
+            is_inside_src_intervals = false
+            for (n, (A,B)) in pairs(f.src_intervals)                                 
+                if (A <= val < B) || (A <= val <= B && n == N)       
+                    newval = rs[n] * val - os[n]                                                                          
+                    out[p] = T <: Integer ? round(Int, newval) : newval  
+                    is_inside_src_intervals = true  
+                    break                                                                                   
+                end                                                                                                    
+            end 
+            if !is_inside_src_intervals
+                # Saturate pixels that fall outside the specified src_intervals by setting
+                # them equal to the start/end of the edge dst_intervals. 
+                a = first(first(f.dst_intervals))
+                b = last(last(f.dst_intervals))
+                A = first(first(f.src_intervals))
+                newval = val < A ? a : b
+                out[p] = T <: Integer ? round(Int, newval) : newval  

```

Comment:
Another option might be to consider using https://github.com/JuliaMath/RoundingIntegers.jl.

Reply:
I don't understand the implication of your suggestion, please could you elaborate? My intention is to handle the case where the user specifies an Integer output type for their output image. Are you suggesting the following?
```julia
out[p] = T <: Integer ? RInt(newval) : newval
```

Reply:
It might not be relevant. I wrote RoundingIntegers because I was a bit uneasy about saying certain algorithms should always feel free to round; what if the user wanted the error instead? So specifying `RInt` makes it possible for the caller to effectively say, "yes, I want rounding" without it being hard-wired into the algorithm.

But using `RInt` has some annoyances for the user, and if you're going to use it "behind the scenes" then I'm not sure it's any better than just calling `round`. So perhaps you can ignore this comment.

Reply:
Meaning, caller could pass in `out = similar(src, RInt)` or something. But that seems picky in this case.